### PR TITLE
Simplify construction of UiParts #184

### LINKED ISSUE #184: Simplify construction of UiParts
Currently there is quite a lot of [boilerplate](https://github.com/se-edu/addressbook-level4/blob/61b1f6fa5d259677efcd8df61b73b7569b7d5715/src/main/java/seedu/address/ui/CommandBox.java#L32-L64) involved if a developer wants to implement a new `UiPart`:

```java
    public static CommandBox load(Stage primaryStage, AnchorPane commandBoxPlaceholder, Logic logic) {
        CommandBox commandBox = UiPartLoader.loadUiPart(primaryStage, commandBoxPlaceholder, new CommandBox());
        commandBox.configure(logic);
        commandBox.addToPlaceholder();
        return commandBox;
    }

    public void configure(Logic logic) {
        this.logic = logic;
        registerAsAnEventHandler(this);
    }

    private void addToPlaceholder() {
        SplitPane.setResizableWithParent(placeHolderPane, false);
        placeHolderPane.getChildren().add(commandTextField);
        FxViewUtil.applyAnchorBoundaryParameters(commandPane, 0.0, 0.0, 0.0, 0.0);
        FxViewUtil.applyAnchorBoundaryParameters(commandTextField, 0.0, 0.0, 0.0, 0.0);
    }

    @Override
    public void setNode(Node node) {
        commandPane = (AnchorPane) node;
    }

    @Override
    public String getFxmlPath() {
        return FXML;
    }

    @Override
    public void setPlaceholder(AnchorPane pane) {
        this.placeHolderPane = pane;
}
```

In short, the developer needs to override:

* `setNode()`
* `setFxmlPath()`
* potentially `setPlaceholder()`

and think of what is the sane thing to do for each of these methods. Additionally, to make the API pleasant to use, the developer also needs to implement a static `load()` method (and in that method, we need to decide which of the three overloaded `UiPartLoader::loadUiPart()` methods we should call)

The reason why is because the current implementation (though `UiPartLoader`) [puts the responsibility on creating new controller instances on the `FXMLLoader`](https://github.com/se-edu/addressbook-level4/blob/master/src/main/java/seedu/address/ui/UiPartLoader.java#L27), but the `FXMLLoader` does not allow us to pass arguments to the constructed controller. This leads to a lot of boilerplate code just to work around this limitation.

By using an approach similar to https://github.com/CS2103AUG2016-T11-C4/main/pull/4 , where we construct the controller *first* and then pass it to the `FXMLLoader`, we can move the FXML loading directly into the `UiPart` constructor and reduce the above code to just:
```java
    public CommandBox(AnchorPane placeHolderPane, Logic logic) {
        super(FXML);
        this.placeHolderPane = placeHolderPane;
        this.logic = logic;
        registerAsAnEventHandler(this);
        addToPlaceholder();
    }

    private void addToPlaceholder() {
        SplitPane.setResizableWithParent(placeHolderPane, false);
        placeHolderPane.getChildren().add(commandTextField);
        FxViewUtil.applyAnchorBoundaryParameters(commandTextField, 0.0, 0.0, 0.0, 0.0);
    }
```
which is much shorter.

cc @yl-coder

### PR DESCRIPTION
**Pull request by pyokagan**
Fixes #184 
Currently there is quite a lot of [boilerplate](https://github.com/se-edu/addressbook-level4/blob/61b1f6fa5d259677efcd8df61b73b7569b7d5715/src/main/java/seedu/address/ui/CommandBox.java#L32-L64) involved if a developer wants to implement a new `UiPart`:

```java
    public static CommandBox load(Stage primaryStage, AnchorPane commandBoxPlaceholder, Logic logic) {
        CommandBox commandBox = UiPartLoader.loadUiPart(primaryStage, commandBoxPlaceholder, new CommandBox());
        commandBox.configure(logic);
        commandBox.addToPlaceholder();
        return commandBox;
    }

    public void configure(Logic logic) {
        this.logic = logic;
        registerAsAnEventHandler(this);
    }

    private void addToPlaceholder() {
        SplitPane.setResizableWithParent(placeHolderPane, false);
        placeHolderPane.getChildren().add(commandTextField);
        FxViewUtil.applyAnchorBoundaryParameters(commandPane, 0.0, 0.0, 0.0, 0.0);
        FxViewUtil.applyAnchorBoundaryParameters(commandTextField, 0.0, 0.0, 0.0, 0.0);
    }

    @Override
    public void setNode(Node node) {
        commandPane = (AnchorPane) node;
    }

    @Override
    public String getFxmlPath() {
        return FXML;
    }

    @Override
    public void setPlaceholder(AnchorPane pane) {
        this.placeHolderPane = pane;
}
```

In short, the developer needs to override:

* `setNode()`
* `setFxmlPath()`
* potentially `setPlaceholder()`

and think of what is the sane thing to do for each of these methods. Additionally, to make the API pleasant to use, the developer also needs to implement a static `load()` method (and in that method, we need to decide which of the three overloaded `UiPartLoader::loadUiPart()` methods we should call)

The reason why is because the current implementation (though `UiPartLoader`) [puts the responsibility on creating new controller instances on the `FXMLLoader`](https://github.com/se-edu/addressbook-level4/blob/master/src/main/java/seedu/address/ui/UiPartLoader.java#L27), but the `FXMLLoader` does not allow us to pass arguments to the constructed controller. This leads to a lot of boilerplate code just to work around this limitation.

With the changes in this PR, we construct the controller *first* and then pass it to the `FXMLLoader`, we can move the FXML loading directly into the `UiPart` constructor and reduce the above code to just:
```java
    public CommandBox(AnchorPane placeHolderPane, Logic logic) {
        super(FXML);
        this.placeHolderPane = placeHolderPane;
        this.logic = logic;
        registerAsAnEventHandler(this);
        addToPlaceholder();
    }

    private void addToPlaceholder() {
        SplitPane.setResizableWithParent(placeHolderPane, false);
        placeHolderPane.getChildren().add(commandTextField);
        FxViewUtil.applyAnchorBoundaryParameters(commandTextField, 0.0, 0.0, 0.0, 0.0);
    }
```
which is much shorter.

### CODE REVIEW

**Comment by canihasreview** [comment]
# v1

@pyokagan submitted v1 for review.

* [v1 1/10] [UiPart: implement FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/62cc0b04109ec74087d998513d427b8e28e5eaa3)
* [v1 2/10] [BrowserPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/cd3398032d474468ebc6983390a40c803da2ba7d)
* [v1 3/10] [PersonListPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/f3f596bb0489dae3858b52e2750234c45f7710a8)
* [v1 4/10] [PersonCard: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/d5af4e9694b45a4532eea8a0fdca7b62f6154619)
* [v1 5/10] [ResultDisplay: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/fe4c30857725fc16f5eb08d2c633b08d8db46b80)
* [v1 6/10] [StatusBarFooter: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/7c1dbf4d813ba777b4570ac2edab865a8fdc7b98)
* [v1 7/10] [CommandBox: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/1b8d5b2137569f9c159a7c478e60fbb67f0ae46a)
* [v1 8/10] [HelpWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/7a8e4bc44465081d8981e602b7ddd0b02212a195)
* [v1 9/10] [MainWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/2ac3a8f1bd4b305929ac49ad496876d28e3cbe92)
* [v1 10/10] [UiPart: remove deprecated construction methods](https://github.com/se-edu/addressbook-level4/pull/237/commits/7f9cd8cd20caf461f7a81b83c5f0f28126e12aab)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d782898510a12b3b03b07848cb7a647f3076e241...7f9cd8cd20caf461f7a81b83c5f0f28126e12aab))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/237/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
Looks like GitHub stopped calling mentionbot web hook. Let me close and reopen to see if that helps.

**Comment by pyokagan** [comment]
@damithc Oh whoops, I just remembered the diagrams also need to be updated as well :-P Will send an updated version.

**Comment by damithc** [comment]
@yl-coder any feedback?

**Comment by canihasreview** [comment]
# v2

@pyokagan submitted v2 for review.

* [v2 1/11] [UiPart: implement FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/750fd9be1edb8d933d872b2faa489885dc8130cc)
* [v2 2/11] [BrowserPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/190d60528f06e9941fb3758be0457557dfa031ef)
* [v2 3/11] [PersonListPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/04fcbfc7365b4da053b0a216c02f6780573b4ec2)
* [v2 4/11] [PersonCard: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/913213eae7f0478f036e7bb460811d95e88a607e)
* [v2 5/11] [ResultDisplay: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/7c45f339f9aadfd7732db977cb0c9993785476ed)
* [v2 6/11] [StatusBarFooter: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/25b42fcebaf00260b3fa9723479b07d4e3a49661)
* [v2 7/11] [CommandBox: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/a97b6eaf40d152f4bce661f2671383f4451d380a)
* [v2 8/11] [HelpWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/126104ae412ed19d2096d02017488e4727df4d57)
* [v2 9/11] [MainWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/f8389d0f7643b7c6e02f21332683461af75e2a4a)
* [v2 10/11] [UiPart: remove deprecated construction methods](https://github.com/se-edu/addressbook-level4/pull/237/commits/52f4831c60dd7468e426a0e8ee1763f75d0cec4c)
* [v2 11/11] [DeveloperGuide: update Ui component info](https://github.com/se-edu/addressbook-level4/pull/237/commits/28327b928d79e814500fe43600d82f0896297376)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/938adb3f65e0680913c63566102f6a887df56445...28327b928d79e814500fe43600d82f0896297376)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/commit/7fc9b3ff76f82664599bc03838664c24421ba9bd))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/237/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by pyokagan** [comment]
Changes since v1:

* Made `UiPart` `abstract` again.

* Updated the developer guide

* `UiPart` test now checks that `@FXML` annotations work.

**Comment by yl-coder** [comment]
Will look at this by tomorrow.

**Comment by CanIHasReview** [comment]
# v3

@pyokagan submitted v3 for review.

* [v3 1/11] [UiPart: implement FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/750fd9be1edb8d933d872b2faa489885dc8130cc)
* [v3 2/11] [BrowserPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/190d60528f06e9941fb3758be0457557dfa031ef)
* [v3 3/11] [PersonListPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/04fcbfc7365b4da053b0a216c02f6780573b4ec2)
* [v3 4/11] [PersonCard: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/913213eae7f0478f036e7bb460811d95e88a607e)
* [v3 5/11] [ResultDisplay: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/7c45f339f9aadfd7732db977cb0c9993785476ed)
* [v3 6/11] [StatusBarFooter: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/25b42fcebaf00260b3fa9723479b07d4e3a49661)
* [v3 7/11] [CommandBox: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/a97b6eaf40d152f4bce661f2671383f4451d380a)
* [v3 8/11] [HelpWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/126104ae412ed19d2096d02017488e4727df4d57)
* [v3 9/11] [MainWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/f8389d0f7643b7c6e02f21332683461af75e2a4a)
* [v3 10/11] [UiPart: remove deprecated construction methods](https://github.com/se-edu/addressbook-level4/pull/237/commits/52f4831c60dd7468e426a0e8ee1763f75d0cec4c)
* [v3 11/11] [DeveloperGuide: update Ui component info](https://github.com/se-edu/addressbook-level4/pull/237/commits/28327b928d79e814500fe43600d82f0896297376)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/938adb3f65e0680913c63566102f6a887df56445...28327b928d79e814500fe43600d82f0896297376)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/bda21dd791f1b6fabff53f829607f70279562810))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/237/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by pyokagan** [comment]
@yl-coder Sorry, please ignore the message from @CanIHasReview . Was just testing.

**Comment by yl-coder** [comment]
Nothing else other than the comments above

**Comment by damithc** [comment]
@yl-coder you can use GitHub's new 'request change' feature when requesting changes.

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 1]
```diff
@@ -60,9 +36,9 @@ private void setConnections(ObservableList<ReadOnlyPerson> personList) {
         setEventHandlerForSelectionChangeEvent();
     }
 
-    private void addToPlaceholder() {
+    private void addToPlaceholder(AnchorPane placeHolderPane) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: @yl-coder 
> Can try to abstract it to its superclass.

`PersonCard` and `HelpWindow` doesn't make use of placeholder panes though. Trying to extract this method to the superclass would place an unnecessary burden on subclasses that do not need placeholder panes.

File: src/main/java/seedu/address/ui/StatusBarFooter.java [hunk 2]
```diff
@@ -16,36 +16,27 @@
...
+        addMainPane(placeHolder);
         setSyncStatus("Not updated yet in this session");
         setSaveLocation("./" + saveLocation);
         registerAsAnEventHandler(this);
     }
 
-    private void addMainPane() {
-        FxViewUtil.applyAnchorBoundaryParameters(mainPane, 0.0, 0.0, 0.0, 0.0);
-        placeHolder.getChildren().add(mainPane);
+    private void addMainPane(AnchorPane placeHolder) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/StatusBarFooter.java]: OK.

File: src/main/resources/view/PersonListPanel.fxml [hunk 3]
```diff
@@ -3,7 +3,7 @@
 <?import javafx.scene.control.ListView?>
 <?import javafx.scene.layout.*?>
 <?import java.net.URL?>
-<VBox xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1" fx:controller="seedu.address.ui.PersonListPanel" AnchorPane.bottomAnchor="0.0" AnchorPane.leftAnchor="0.0" AnchorPane.rightAnchor="0.0" AnchorPane.topAnchor="0.0">
+<VBox xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1" AnchorPane.bottomAnchor="0.0" AnchorPane.leftAnchor="0.0" AnchorPane.rightAnchor="0.0" AnchorPane.topAnchor="0.0">
```
Comment (pyokagan) [inline src/main/resources/view/PersonListPanel.fxml]: OK, I've added that as a cleanup commit.

**Comment by CanIHasReview** [comment]
# v3

@pyokagan submitted v3 for review.

* [v3 1/13] [UiPart: implement FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/549e2ce6c770a17bdacfa82e2160b02853629f92)
* [v3 2/13] [BrowserPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/1da3e359b65879d0fde3b1c9849441087e1883d1)
* [v3 3/13] [PersonListPanel: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/16ba2f9b9c6fe2813ba3f52b17d45190fea7a2cb)
* [v3 4/13] [PersonCard: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/443d6fa012adeffad8e7448e1d40d14fe15ce7c5)
* [v3 5/13] [ResultDisplay: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/4b229e3ada91eed6af32db7635ecb229c617d5ce)
* [v3 6/13] [StatusBarFooter: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/e78d245a577cbe301bf59e2fd1e0a6d006f6345d)
* [v3 7/13] [CommandBox: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/67be0359491387785849f434c4fd4e45006e47e3)
* [v3 8/13] [HelpWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/dc69c618a1470a5efae330fb17b234e3e9cf6877)
* [v3 9/13] [MainWindow: migrate to use FXML loading constructor](https://github.com/se-edu/addressbook-level4/pull/237/commits/0c7cc550f88f6b21481e4c078dc0d08c40cbc834)
* [v3 10/13] [UiPart: remove deprecated construction methods](https://github.com/se-edu/addressbook-level4/pull/237/commits/fa7dbdef60152365271b7fb2231fa1e97f266ca8)
* [v3 11/13] [DeveloperGuide: update Ui component info](https://github.com/se-edu/addressbook-level4/pull/237/commits/2cb1e4ef7672b56e4a7599b1306f5970d8cffef6)
* [v3 12/13] [StatusBarFooter: rename addMainPane() to addToPlaceholder()](https://github.com/se-edu/addressbook-level4/pull/237/commits/88b4c9abbd4687bf59544a4ef9b79b4eb9c86108)
* [v3 13/13] [PersonListPanel: use FxViewUtil.applyAnchorBoundaryParameters()](https://github.com/se-edu/addressbook-level4/pull/237/commits/8711079d66e3cfd51e2835ecd9728cfc3d0c2719)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/0dd249a002f77bde21542f679f19f3d4bfa0444c...8711079d66e3cfd51e2835ecd9728cfc3d0c2719)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/482dca5f2d093dc69d67fb4603178b9898f58261))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/237/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by pyokagan** [comment]
Changes from v2..v3:

* [v3 12/13] and [v3 13/13] are new commits.

* Corrected spelling errors in the commit messages.

**Comment by pyokagan** [comment]
@damithc Can I confirm your approval? The PR has changed since your last review (v1).

**Pull request merged**